// @taiga-ui/kit/esm2022/components/tiles/tile-handle.directive.mjs
import { __decorate } from "tslib";
import { Directive, inject } from '@angular/core';
import { tuiGetActualTarget, tuiIsElement } from '@taiga-ui/cdk/utils/dom';
import { shouldCall } from '@taiga-ui/event-plugins';
import { TuiTile } from './tile.component';
import * as i0 from "@angular/core";
function isInteracting(x = NaN) {
    return !Number.isNaN(x) || !Number.isNaN(this['x']);
}
function isDragging() {
    return !Number.isNaN(this['x']);
}
class TuiTileHandle {
    constructor() {
        this.tile = inject(TuiTile);
        this.x = NaN;
        this.y = NaN;
    }
    onPointer(x = NaN, y = NaN) {
        const { left, top } = this.tile.element.getBoundingClientRect();
        this.x = x - left;
        this.y = y - top;
        this.tile.onDrag([NaN, NaN]);
    }
    onMove(x, y) {
        this.tile.onDrag([x - this.x, y - this.y]);
    }
    onStart(event) {
        const target = tuiGetActualTarget(event);
        const { x, y, pointerId } = event;
        if (tuiIsElement(target)) {
            target.releasePointerCapture(pointerId);
        }
        this.onPointer(x, y);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTileHandle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTileHandle, isStandalone: true, selector: "[tuiTileHandle]", host: { listeners: { "pointerdown.silent.prevent": "onStart($event)", "document:pointerup.silent": "onPointer()", "document:pointermove.silent": "onMove($event.x, $event.y)" }, properties: { "style.touchAction": "\"none\"", "style.userSelect": "\"none\"" } }, ngImport: i0 }); }
}
__decorate([
    shouldCall(isInteracting)
], TuiTileHandle.prototype, "onPointer", null);
__decorate([
    shouldCall(isDragging)
], TuiTileHandle.prototype, "onMove", null);
export { TuiTileHandle };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTileHandle, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTileHandle]',
                    host: {
                        '[style.touchAction]': '"none"',
                        '[style.userSelect]': '"none"',
                        '(pointerdown.silent.prevent)': 'onStart($event)',
                        '(document:pointerup.silent)': 'onPointer()',
                        '(document:pointermove.silent)': 'onMove($event.x, $event.y)',
                    },
                }]
        }], propDecorators: { onPointer: [], onMove: [] } });
