// @taiga-ui/core/esm2022/directives/popup/popup.directive.mjs
import { Directive, inject, Input, TemplateRef } from '@angular/core';
import { TuiPopupService } from './popup.service';
import * as i0 from "@angular/core";
class TuiPopup {
    constructor() {
        this.template = inject(TemplateRef);
        this.service = inject(TuiPopupService);
    }
    set tuiPopup(show) {
        this.viewRef?.destroy();
        if (show) {
            this.viewRef = this.service.addTemplate(this.template);
        }
    }
    ngOnDestroy() {
        this.viewRef?.destroy();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopup, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPopup, isStandalone: true, selector: "ng-template[tuiPopup]", inputs: { tuiPopup: "tuiPopup" }, ngImport: i0 }); }
}
export { TuiPopup };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopup, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiPopup]',
                }]
        }], propDecorators: { tuiPopup: [{
                type: Input
            }] } });
