// @taiga-ui/addon-mobile/esm2022/directives/responsive-dialog/responsive-dialog.service.mjs
import { inject, Injectable } from '@angular/core';
import { TuiSheetDialogService } from '@taiga-ui/addon-mobile/components/sheet-dialog';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiIsString } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiDialogService } from '@taiga-ui/core/components/dialog';
import * as i0 from "@angular/core";
/**
 * Same as `TuiDialogService` but automatically switches to `TuiSheetDialogService` on mobile
 */
class TuiResponsiveDialogService {
    constructor() {
        this.isMobile = inject(TUI_IS_MOBILE);
        this.dialogs = inject(TuiDialogService);
        this.sheets = inject(TuiSheetDialogService);
    }
    open(content, options = {}) {
        return this.isMobile
            ? this.sheets.open(content, options)
            : this.dialogs.open(content, {
                ...options,
                label: tuiIsString(options.label) ? options.label : '',
            });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiResponsiveDialogService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiResponsiveDialogService, providedIn: 'root' }); }
}
export { TuiResponsiveDialogService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiResponsiveDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
